### DISCUSSION TITLE
Skip "a target while its command is running if a condition is met"

### DISCUSSION THREAD

**Discussion by edalfon** [ropensci/targets#847] [Help]
Is there any way to skip a target while its command is running?

As you can see from the description, I am looking for pretty much the same behavior as `tar_cancel`, but instead of showing the target as cancelled, I wanted it to treat the target as skipped (so that the info -e.g. running time- from previous run remains the same, it appears and up to date, and also downstream targets that depend on it also appear as up to date).

I am aware of `tar_cue`, `tar_cue_skip` and friends, that you can use to customize the cue argument of a target. But I want to decide to skip the target at running time, and not at the time of the definition of the pipeline.

Also there's `tar_skip`, that let's you create a target that cancel itself. But could not find anything to create a target that skips itself.

**Comment by wlandau** [comment]
When you cancel a target `x` with `tar_cancel()`, the metadata from the previous run is left alone. So for runtimes specifically, `tar_meta(x, fields = all_of("seconds"))` reflects the last completed run. At that point, `tar_outdated()` may show `x` and downstream targets as outdated. However, since the hash of `x` does not change if you cancel it, the behavior of `x` will not actually cause downstream targets to rerun.

A full skip may weaken the reproducibility guarantees of a `targets` pipeline, and it is not the sort of thing I would encourage.

**Reply by edalfon** [reply]
Thanks for the quick feedback.

> When you cancel a target `x` with `tar_cancel()`, the metadata from the previous run is left alone. So for runtimes specifically, `tar_meta(x, fields = all_of("seconds"))` reflects the last completed run. At that point, `tar_outdated()` may show `x` and downstream targets as outdated. However, since the hash of `x` does not change if you cancel it, the behavior of `x` will not actually cause downstream targets to rerun.

Yeah. Perhaps I forgot to explain that in terms of running the pipeline, `tar_cancel()` does exactly what I was looking for. Because as you explain, downstream targets will not re-run. But then inspecting the pipeline becomes troublesome, because all downstream targets will show as outdated, as long as the upstream target keeps cancelled.

That was the issue that triggered my question. I had a pipeline and in one target at the very beginning we checked a condition and cancelled it. Then all the downstream targets (the hole pipeline) keep showing as outdated and you cannot really tell when it's because of other upstream targets, other than the cancelled one. 

> A full skip may weaken the reproducibility guarantees of a `targets` pipeline, and it is not the sort of thing I would encourage.

Agree, it may weaken reproducibility guarantees. But it may depend on what cancelled/skipped means for the target. For example, if it means that it could not connect to the database to get the latest data, just skipping it can be an issue because it will not run again. But if it's something that the target itself checks and finds the result will not change, it may be ok.

I have in mind a case that looks something like this:
- There's a target that need to re-run whenever a file changes.
- Downstream targets operate on that file.
- But the result of some of the downstream targets will not change, even with the new data.
- It is cheap (in terms of time) to check if the data changes will affect the result of the target. But it's expensive to do the actual calculations.
- So the idea was to test this at the beginning of the target and skip its execution accordingly. But this would not be a failed run, so the target would not need to re-run unless other dependencies invalidates it again. Nor would need to run downstream targets that depends on it.

`tar_cancel` achieves almost this, but the target keeps marked as cancelled and all the downstream targets as outdated.

An alternative could be to try and return early the value of the target so that it successfully completes execution, but then we would lose in the pipeline the information (running time) of the actual calculations.

Or perhaps we should re-think the structure of the pipeline and choose better targets ([although we already keep in mind your advice on choosing good targets](https://books.ropensci.org/targets/targets.html)).

Any ideas or examples on how to deal with similar cases?

**Reply by wlandau** [reply]
> It is cheap (in terms of time) to check if the data changes will affect the result of the target. But it's expensive to do the actual calculations.

In that case, you could break up that target into two different targets: one for stuff that happens before `tar_cancel()`, and everything else that happens after `tar_cancel()`. Then you wouldn't need `tar_cancel()` at all. I didn't want to inundate https://books.ropensci.org/targets/targets.html with this, but if different pieces of work have different invalidation criteria (i.e. when is a step "up to date"?) then that can be a good reason to break up a target into multiple targets.
